File: src/main/kotlin/io/engrate/powertariffs/importers/MeteringGridAreasImporter.kt
package io.engrate.powertariffs.importers

import io.engrate.powertariffs.persistence.repository.GridOperatorsRepository
import io.engrate.powertariffs.persistence.repository.MeteringGridAreasRepository
import io.engrate.powertariffs.persistence.schema.MeteringGridArea
import jakarta.transaction.Transactional
import org.slf4j.LoggerFactory
import org.springframework.boot.ApplicationRunner
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.core.annotation.Order
import org.springframework.core.io.ClassPathResource
import org.springframework.stereotype.Component

@Order(2)
@Component
@Transactional
@EnableConfigurationProperties(ImportersProperties::class)
class MeteringGridAreasImporter(
    private val meteringGridAreaRepository: MeteringGridAreasRepository,
    private val gridOperatorsRepository: GridOperatorsRepository,
    private val props: ImportersProperties,
) : ApplicationRunner {
    private val logger = LoggerFactory.getLogger(MeteringGridAreasImporter::class.java)

    override fun run(args: org.springframework.boot.ApplicationArguments?) {
        if (!props.loadMeteringGridAreas) {
            logger.info("Skipping metering grid areas import")
            return
        }
        val csvPath = "data/metering_grid_areas/mgas.csv"
        val resource = ClassPathResource(csvPath)
        if (!resource.exists()) {
            logger.error("Metering grid areas CSV file not found at: $csvPath")
            return
        }

        resource.inputStream.bufferedReader(Charsets.UTF_8).use { reader ->
            val lines = reader.readLines()
            if (lines.isEmpty()) {
                logger.warn("Metering grid areas CSV is empty: $csvPath")
                return@use
            }

            val rows = lines.drop(1) // No header check, skip first line
            for (rowLine in rows) {
                val row = rowLine.split(";")
                if (row.size < 6) {
                    logger.warn("Skipping malformed row: $row")
                    continue
                }

                val operatorName = row[0].trim()
                val mgaName = row[1].trim()
                val mgaCode = row[2].trim()
                val mgaType = row[3].trim()
                val mbaCode = row[4].trim()
                val country = row[5].trim()

                if (!mgaType.contains("DISTRIBUTION", ignoreCase = true)) {
                    logger.warn("Skipping $mgaName - only DISTRIBUTION types are supported")
                    continue
                }

                val operator = gridOperatorsRepository.findByName(operatorName)
                if (operator == null) {
                    logger.warn("Operator $operatorName does not exist, skipping")
                    continue
                }

                val existingMga = meteringGridAreaRepository.findByCode(mgaCode)
                if (existingMga != null) {
                    logger.warn("Metering grid area $mgaCode already exists, skipping")
                    continue
                }

                val mga =
                    MeteringGridArea(
                        code = mgaCode,
                        name = mgaName,
                        countryCode = country,
                        meteringBusinessArea = mbaCode,
                        gridOperator = operator,
                        tariffAssociations = emptySet(),
                    )

                meteringGridAreaRepository.save(mga)
                logger.info("Metering grid area saved: $mgaName (Code: $mgaCode)")
            }
        }
    }
}

File: src/importers/metering_grid_areas_importer.py
from engrate_sdk.utils import log

from src.model import MeteringGridAreaSpec
from src.repositories.power_tariffs_repository import repository

logger = log.get_logger(__name__)


async def load_metering_grid_areas():
    from pathlib import Path
    import csv

    with open(Path("data/metering_grid_areas/mgas.csv"), "r", encoding="utf-8") as file:
        reader = csv.reader(file, delimiter=";")
        for row in reader:
            if len(row) < 6:
                logger.warning(f"Skipping malformed row: {row}")
                continue

            # Extract fields
            operator_name = row[0].strip()
            mga_name = row[1].strip()
            mga_code = row[2].strip()
            mga_type = row[3].strip()
            mba_code = row[4].strip()
            country = row[5].strip()

            # Skip if not a grid operator (Nätägare)
            if "DISTRIBUTION" not in mga_type:
                logger.warning(
                    f"Skipping {mga_name} - only DISTRIBUTION types are supported"
                )
                continue

            # Check if operator already exists
            existing_operator = await repository.get_operator_by_name(operator_name)
            if not existing_operator:
                logger.warning(f"Operator {operator_name} does not exists, skipping")
                continue

            mag = MeteringGridAreaSpec(
                code=mga_code,
                country_code=country,
                name=mga_name,
                metering_business_area=mba_code,
                grid_operator=existing_operator.model_dump(),
            )
            existing_mga = await repository.get_metering_grid_area_by_code(mga_code)
            if existing_mga:
                logger.warning(
                    f"Metering grid area {mga_code} already exists, skipping"
                )
                continue
            await repository.save_metering_grid_area(mag)
            logger.info(f"Metering grid area saved: {mga_name} (Code: {mga_code})")
